File: samples/tutorials-xamarin/tutorials-3d/tutorial-03/MainActivity.cs
using System;
using Android.App;
using Android.Graphics;
using Android.OS;
using Android.Views;
using AndroidX.AppCompat.App;
using SciChart.Charting.Visuals;
using SciChart.Charting3D.Model.DataSeries.Xyz;
using SciChart.Charting3D.Modifiers;
using SciChart.Charting3D.Visuals;
using SciChart.Charting3D.Visuals.Axes;
using SciChart.Charting3D.Visuals.PointMarkers;
using SciChart.Charting3D.Visuals.RenderableSeries.Scatter;

namespace Tutorial_3d_03
{
    [Activity(Label = "@string/app_name", Theme = "@style/AppTheme.NoActionBar", MainLauncher = true)]
    public class MainActivity : AppCompatActivity
    {
        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            Xamarin.Essentials.Platform.Init(this, savedInstanceState);

            SciChartSurface.SetRuntimeLicenseKey("");
            SetContentView(Resource.Layout.activity_main);

            var surface = new SciChartSurface3D(this);
            var chartLayout = FindViewById<ViewGroup>(Resource.Id.chart_layout);
            chartLayout.AddView(surface);

            var dataSeries = new XyzDataSeries3D<double, double, double>();
            for (int i = 0; i < 200; i++)
            {
                var x = GetGaussianRandomNumber(5.0, 1.5);
                var y = GetGaussianRandomNumber(5.0, 1.5);
                var z = GetGaussianRandomNumber(5.0, 1.5);
                dataSeries.Append(x, y, z);
            }

            var rSeries = new ScatterRenderableSeries3D()
            {
                DataSeries = dataSeries,
                PointMarker = new SpherePointMarker3D()
                {
                    Fill = new Color(0x32, 0xCD, 0x32, 0xFF),
                    Size = 10
                }
            };

            // <CreateTooltipModifier>
            var tooltipModifier = new TooltipModifier3D { CrosshairMode = CrosshairMode.Lines };
            // </CreateTooltipModifier>

            using (surface.SuspendUpdates())
            {
                surface.XAxis = new NumericAxis3D();
                surface.YAxis = new NumericAxis3D();
                surface.ZAxis = new NumericAxis3D();

                surface.RenderableSeries.Add(rSeries);

                // <AddModifiers>
                surface.ChartModifiers.Add(new OrbitModifier3D { ExecuteOnPointerCount = 2 });
                surface.ChartModifiers.Add(new ZoomExtentsModifier3D());
                surface.ChartModifiers.Add(new PinchZoomModifier3D());
                surface.ChartModifiers.Add(tooltipModifier);
                // </AddModifiers>
            }
        }   

        private double GetGaussianRandomNumber(double mean, double stdDev)
        {
            var random = new Random();

            double u1 = random.NextDouble();
            double u2 = random.NextDouble();

            var randStdNormal = Math.Sqrt(-2.0 * Math.Log10(u1)) * Math.Sin(2.0 * Math.PI * u2);
            return mean * stdDev * randStdNormal;
        }
    }
}

File: samples/tutorials-native/tutorials-3d/tutorial-3/javaBuilder/src/main/java/com/scichart/tutorial/MainActivity.java
package com.scichart.tutorial;

import android.os.Bundle;
import android.util.Log;
import android.widget.LinearLayout;

import androidx.appcompat.app.AppCompatActivity;

import com.scichart.charting.visuals.SciChartSurface;
import com.scichart.charting3d.model.dataSeries.xyz.XyzDataSeries3D;
import com.scichart.charting3d.modifiers.OrbitModifier3D;
import com.scichart.charting3d.modifiers.PinchZoomModifier3D;
import com.scichart.charting3d.modifiers.TooltipModifier3D;
import com.scichart.charting3d.modifiers.ZoomExtentsModifier3D;
import com.scichart.charting3d.visuals.SciChartSurface3D;
import com.scichart.charting3d.visuals.pointMarkers.SpherePointMarker3D;
import com.scichart.charting3d.visuals.renderableSeries.scatter.ScatterRenderableSeries3D;
import com.scichart.core.framework.UpdateSuspender;
import com.scichart.extensions3d.builders.SciChart3DBuilder;

import java.util.Collections;
import java.util.Random;

public class MainActivity extends AppCompatActivity {

    private SciChartSurface3D surface;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        try {
            SciChartSurface.setRuntimeLicenseKey("");
        } catch (Exception e) {
            Log.e("SciChart", "Error during setting license key", e);
        }

        LinearLayout chartLayout = findViewById(R.id.chart_layout);

        surface = new SciChartSurface3D(this);
        chartLayout.addView(surface);

        SciChart3DBuilder.init(this);
        final SciChart3DBuilder sciChartBuilder = SciChart3DBuilder.instance();

        final XyzDataSeries3D dataSeries = new XyzDataSeries3D(Double.class, Double.class, Double.class);
        for (int i = 0; i < 200; i++) {
            final double x = getGaussianRandomNumber(5.0, 1.5);
            final double y = getGaussianRandomNumber(5.0, 1.5);
            final double z = getGaussianRandomNumber(5.0, 1.5);
            dataSeries.append(x, y, z);
        }

        final SpherePointMarker3D pointMarker = sciChartBuilder.newSpherePointMarker3D()
                .withFill(0xFF32CD32)
                .withSize(10.0f)
                .build();

        final ScatterRenderableSeries3D rSeries = sciChartBuilder.newScatterSeries3D()
                .withDataSeries(dataSeries)
                .withPointMarker(pointMarker)
                .build();

        // <CreateTooltipModifier>
        final TooltipModifier3D tooltipModifier = new TooltipModifier3D();
        tooltipModifier.setExecuteOnPointerCount(2);
        // </CreateTooltipModifier>

        UpdateSuspender.using(surface, () -> {
            surface.setXAxis(sciChartBuilder.newNumericAxis3D().build());
            surface.setYAxis(sciChartBuilder.newNumericAxis3D().build());
            surface.setZAxis(sciChartBuilder.newNumericAxis3D().build());
            Collections.addAll(surface.getRenderableSeries(), rSeries);

            // <AddModifiers>
            Collections.addAll(surface.getChartModifiers(),
                    new OrbitModifier3D(),
                    new ZoomExtentsModifier3D(),
                    new PinchZoomModifier3D(),
                    tooltipModifier
            );
            // </AddModifiers>
        });
    }

    private Double getGaussianRandomNumber(Double mean, Double stdDev) {
        final Random random = new Random();

        final double u1 = random.nextDouble();
        final double u2 = random.nextDouble();

        final double randStdNormal = Math.sqrt(-2.0 * Math.log(u1)) * Math.sin(2.0 * Math.PI * u2);
        return mean * stdDev * randStdNormal;
    }
}
